util/unifyfs-stage/src/unifyfs-stage-transfer.c

```diff
@@ -0,0 +1,294 @@
+/*
+ * Copyright (c) 2020, Lawrence Livermore National Security, LLC.
+ * Produced at the Lawrence Livermore National Laboratory.
+ *
+ * Copyright 2020, UT-Battelle, LLC.
+ *
+ * LLNL-CODE-741539
+ * All rights reserved.
+ *
+ * This is the license for UnifyFS.
+ * For details, see https://github.com/LLNL/UnifyFS.
+ * Please read https://github.com/LLNL/UnifyFS/LICENSE for full license text.
+ */
+#include <config.h>
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <stdint.h>
+#include <string.h>
+#include <errno.h>
+#include <ctype.h>
+#include <limits.h>
+#include <sys/time.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+#include <unistd.h>
+#include <fcntl.h>
+#include <dirent.h>
+#include <libgen.h>
+#include <getopt.h>
+#include <time.h>
+#include <mpi.h>
+#include <openssl/md5.h>
+
+#include "unifyfs-stage.h"
+
+static int md5_checksum(const char* path, unsigned char* digest)
+{
+    int ret = 0;
+    size_t len = 0;
+    int fd = -1;
+    unsigned char data[1024] = { 0, };
+    MD5_CTX md5;
+
+    fd = open(path, O_RDONLY);
+    if (fd < 0) {
+        perror("open");
+        return errno;
+    }
+
+    ret = MD5_Init(&md5);
+    if (!ret) {
+        fprintf(stderr, "failed to create md5 context\n");
+        goto out;
+    }
+
+    while ((len = read(fd, (void*) data, 1024)) != 0) {
+        ret = MD5_Update(&md5, data, len);
+        if (!ret) {
+            fprintf(stderr, "failed to update checksum\n");
+            goto out;
+        }
+    }
+
+    ret = MD5_Final(digest, &md5);
+    if (!ret) {
+        fprintf(stderr, "failed to finalize md5\n");
+    }
+
+out:
+    /* MD5_xx returns 1 for success */
+    ret = ret == 1 ? 0 : EIO;
+    close(fd);
+
+    return ret;
+}
+
+static inline char* checksum_str(char* buf, unsigned char* digest)
+{
+    int i = 0;
+    char* pos = buf;
+
+    for (i = 0; i < MD5_DIGEST_LENGTH; i++) {
+        pos += sprintf(pos, "%02x", digest[i]);
+    }
+
+    pos[0] = '\0';
+
+    return buf;
+}
+
+static int verify_checksum(const char* src, const char* dst)
+{
+    int ret = 0;
+    int i = 0;
+    char md5src[36] = { 0, };
```

Comment:
hard-coded 36 here is bound to cause problems. use UNIFYFS_MAX_FILENAME for all path buffers